src/host/linalg.jl

```diff
@@ -127,6 +127,35 @@ else
     end
 end
 
+Base.:\(D::Diagonal, b::AbstractGPUVector) = typeof(b)(D.diag) .\ b
+Base.:\(D::Diagonal, B::AbstractGPUMatrix) = typeof(B)(hcat(D.diag))[:, 1] .\ B
```

Comment:
Not sure about this construction. In CUDA.jl this would be 
```julia
Base.:\(D::Diagonal, B::AnyCuMatrix) = cu(D.diag) .\ B
```
I forget where, but wasn't there some discussion about providing a device-independent way of constructing a close representative of an `AbstractArray` on the GPU?

Reply:
That would possibly trigger an upload, which we generally don't want. So I think it would be better to use `Diagonal{..., <:AbstractGPUArray}`. The memory is then on the GPU already and doesn't need a call to the constructor (which here would materialize the diagonal, which isn't something we want?)

Reply:
Ah, that's less general then what I had in mind. When the first argument isn't a GPU array, what shall we do? We can download he second argument and call the CPU method, or we can upload the first argument (which is much lighter) and perform the operation on the GPU, which is generally faster (what I opted for here). Third option would be to not implement this, which will probably default to the method in LinearAlgebra which triggers scalar indexing.

Reply:
> When the first argument isn't a GPU array, what shall we do?

It's the user's responsibility to make sure that doesn't happen. There isn't a good answer as to what to do (upload or download), so we don't provide such methods. You also can't broadcast a CPU and a GPU array.

I realize that's not the most user-friendly option though, so maybe we should start adding useful error messages when this situation occurs. On the other hand, scalar indexing is there for exactly that purpose, with the added benefit that the code still works (i.e. doesn't error).

Reply:
> It's the user's responsibility to make sure that doesn't happen.
That's one possibility I suppose. I'm coming at this from an angle of trying to make generic algorithms work _and_ be reasonably performant on the GPU out of the box. This might require a few opinionated default choices. But maybe this is something for a different package? I just find that most of the time, if part of the data I'm working on is already on the GPU, and I want to do some operation which is generally fast on the GPU, I want to also upload the other data. The converse hasn't happened to me quite so often.

So you would vouch to restrict this method to accept only GPU arrays in the first argument?

Reply:
It's not generic though, you're passing a CPU array into a GPU function. For consistency, I'd like to stick with the current behavior for now, even more because your upload here materializes the diagonal Just do `D.diag .\ b` and have `broadcast` error when a CPU array is involved, there's no need to constrain the first argument.

Reply:
> It's not generic though, you're passing a CPU array into a GPU function.

You mean because this is a GPU package? `Base.:\` by itself is generic though, and from a very detached user perspective it does not really matter whether its arguments live on different devices, particularly if this operation is nested very deep in some application. It would be great if one could just pass a GPU input and this would get propagated through the code for all operations which profit from this (which is something that often generalized across different applications, e.g. linear algebra is usually faster on the GPU).

Maybe one way of addressing this would be a function `default_upload` or `default_download`, similar in spirit to `CUDA.allowscalar`.

But I do get your argument in favor of cleanliness.
I'll implement your suggestions.

Reply:
> You mean because this is a GPU package?

I meant that the calling code probably isn't generic because it constructed a CPU Diagonal next to a GPU Array. Depending on where that Diagonal came from, it probably should have been constructed using a `similar` call to copy the other array its properties. Isn't that possible in a generic way, without special-casing for CuArray?

Furthermore, by calling the GPU array constructor here, you may lose structure (`.diag` could be a SubArray or so).